pkg/webhook/v1alpha2/applicationconfiguration/mutating_handler.go

```diff
@@ -0,0 +1,174 @@
+
+	marshalled, err := json.Marshal(obj)
+	if err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+
+	resp := admission.PatchResponseFromRaw(req.AdmissionRequest.Object.Raw, marshalled)
+	if len(resp.Patches) > 0 {
+		mutatelog.Info("admit ApplicationConfiguration",
+			"namespace", obj.Namespace, "name", obj.Name, "patches", util.JSONMarshal(resp.Patches))
+	}
+	return resp
+}
+
+// Mutate sets all the default value for the Component
+func (h *MutatingHandler) Mutate(obj *v1alpha2.ApplicationConfiguration) error {
+	mutatelog.Info("mutate", "name", obj.Name)
+
+	for compIdx, comp := range obj.Spec.Components {
+		var updated bool
+		for idx, tr := range comp.Traits {
+			var content map[string]interface{}
+			if err := json.Unmarshal(tr.Trait.Raw, &content); err != nil {
+				return err
+			}
+			rawByte, mutated, err := h.mutateTrait(content, comp.ComponentName, obj.GetNamespace())
+			if err != nil {
+				return err
+			}
+			if !mutated {
+				continue
+			}
+			tr.Trait.Raw = rawByte
+			comp.Traits[idx] = tr
+			updated = true
+		}
+		if updated {
+			obj.Spec.Components[compIdx] = comp
+		}
+	}
+
+	return nil
+}
+
+func (h *MutatingHandler) mutateTrait(content map[string]interface{}, compName, namespace string) ([]byte, bool, error) {
+	if content[TraitTypeField] == nil {
+		return nil, false, nil
+	}
+	traitType, ok := content[TraitTypeField].(string)
+	if !ok {
+		return nil, false, fmt.Errorf("name of trait should be string instead of %s", reflect.TypeOf(content[TraitTypeField]))
+	}
+	mutatelog.Info("the trait refers to traitDefinition by name", "compName", compName, "trait name", traitType)
+	// Fetch the corresponding traitDefinition CR, the traitDefinition crd is cluster scoped
+	traitDefinition := &v1alpha2.TraitDefinition{}
+	if err := h.Client.Get(context.TODO(), types.NamespacedName{Name: traitType}, traitDefinition); err != nil {
+		return nil, false, err
+	}
+	// fetch the CRDs definition
+	customResourceDefinition := &crdv1.CustomResourceDefinition{}
+	if err := h.Client.Get(context.TODO(), types.NamespacedName{Name: traitDefinition.Spec.Reference.Name}, customResourceDefinition); err != nil {
+		return nil, false, err
+	}
+	// reconstruct the trait CR
+	delete(content, TraitTypeField)
+
+	if content[TraitSpecField] != nil {
+		content["spec"] = content[TraitSpecField]
+		delete(content, TraitSpecField)
+	}
+
+	trait := unstructured.Unstructured{
+		Object: content,
+	}
+	// find out the GVK from the CRD definition and set
+	apiVersion := metav1.GroupVersion{
+		Group:   customResourceDefinition.Spec.Group,
+		Version: customResourceDefinition.Spec.Versions[0].Name,
+	}.String()
+	trait.SetAPIVersion(apiVersion)
+	trait.SetKind(customResourceDefinition.Spec.Names.Kind)
+	mutatelog.Info("Set the trait GVK", "trait api version", trait.GetAPIVersion(), "trait Kind", trait.GetKind())
+	// copy namespace/label/annotation to the trait and add traitType label
+	trait.SetNamespace(namespace)
```

Comment:
why is the namespace needed to be copied? Is it not in the content before?

Reply:
yes， I shouldn't add it here, thanks, will fix it.